packages/block-editor/src/components/use-block-commands/index.js

```diff
@@ -41,6 +41,21 @@ export const useTransformCommands = () => {
 			const selectedBlocks = getBlocksByClientId(
 				selectedBlockClientIds
 			);
+
+			// selectedBlocks can have `null`s when something tries to call `selectBlock` with an inexistent clientId.
+			// These nulls will cause fatal errors down the line.
+			// In order to prevent discrepancies between selectedBlockClientIds and selectedBlocks, we effectively treat the entire selection as invalid.
+			// We return a set of empty arrays to indicate that nothing is really selected and no transforms can be performed on this 'illegal' selection.
+			// @see https://github.com/WordPress/gutenberg/pull/59410#issuecomment-2006304536
+			if ( selectedBlocks.filter( ( block ) => ! block ).length > 0 ) {
+				return {
+					blocks: [],
+					clientIds: [],
+					possibleBlockTransformations: [],
```

Comment:
This will return a new array reference on each run and trick shallow equality check that the store has "fresh" data. We could use a stable `EMPTY_ARRAY` reference, but it might be easier just to return a flag value here and then bail out early in the command.


Here's an example patch (sorry for whitespace changes).

<details><summary>Details</summary>
<p>

```diff
diff --git packages/block-editor/src/components/use-block-commands/index.js packages/block-editor/src/components/use-block-commands/index.js
index 489a2f602d5..8d608682b2a 100644
--- packages/block-editor/src/components/use-block-commands/index.js
+++ packages/block-editor/src/components/use-block-commands/index.js
@@ -27,51 +27,50 @@ import { store as blockEditorStore } from '../../store';
 
 export const useTransformCommands = () => {
 	const { replaceBlocks, multiSelect } = useDispatch( blockEditorStore );
-	const { blocks, clientIds, canRemove, possibleBlockTransformations } =
-		useSelect( ( select ) => {
-			const {
-				getBlockRootClientId,
-				getBlockTransformItems,
-				getSelectedBlockClientIds,
-				getBlocksByClientId,
-				canRemoveBlocks,
-			} = select( blockEditorStore );
-
-			const selectedBlockClientIds = getSelectedBlockClientIds();
-			const selectedBlocks = getBlocksByClientId(
-				selectedBlockClientIds
-			);
+	const {
+		blocks,
+		clientIds,
+		canRemove,
+		possibleBlockTransformations,
+		invalidBlocks,
+	} = useSelect( ( select ) => {
+		const {
+			getBlockRootClientId,
+			getBlockTransformItems,
+			getSelectedBlockClientIds,
+			getBlocksByClientId,
+			canRemoveBlocks,
+		} = select( blockEditorStore );
 
-			// selectedBlocks can have `null`s when something tries to call `selectBlock` with an inexistent clientId.
-			// These nulls will cause fatal errors down the line.
-			// In order to prevent discrepancies between selectedBlockClientIds and selectedBlocks, we effectively treat the entire selection as invalid.
-			// We return a set of empty arrays to indicate that nothing is really selected and no transforms can be performed on this 'illegal' selection.
-			// @see https://github.com/WordPress/gutenberg/pull/59410#issuecomment-2006304536
-			if ( selectedBlocks.filter( ( block ) => ! block ).length > 0 ) {
-				return {
-					blocks: [],
-					clientIds: [],
-					possibleBlockTransformations: [],
-					canRemove: false,
-				};
-			}
+		const selectedBlockClientIds = getSelectedBlockClientIds();
+		const selectedBlocks = getBlocksByClientId( selectedBlockClientIds );
+
+		// selectedBlocks can have `null`s when something tries to call `selectBlock` with an inexistent clientId.
+		// These nulls will cause fatal errors down the line.
+		// In order to prevent discrepancies between selectedBlockClientIds and selectedBlocks, we effectively treat the entire selection as invalid.
+		// We return a set of empty arrays to indicate that nothing is really selected and no transforms can be performed on this 'illegal' selection.
+		// @see https://github.com/WordPress/gutenberg/pull/59410#issuecomment-2006304536
+		if ( selectedBlocks.filter( ( block ) => ! block ).length > 0 ) {
+			return { invalidBlocks: true };
+		}
 
-			const rootClientId = getBlockRootClientId(
-				selectedBlockClientIds[ 0 ]
-			);
-			return {
-				blocks: selectedBlocks,
-				clientIds: selectedBlockClientIds,
-				possibleBlockTransformations: getBlockTransformItems(
-					selectedBlocks,
-					rootClientId
-				),
-				canRemove: canRemoveBlocks(
-					selectedBlockClientIds,
-					rootClientId
-				),
-			};
-		}, [] );
+		const rootClientId = getBlockRootClientId(
+			selectedBlockClientIds[ 0 ]
+		);
+		return {
+			blocks: selectedBlocks,
+			clientIds: selectedBlockClientIds,
+			possibleBlockTransformations: getBlockTransformItems(
+				selectedBlocks,
+				rootClientId
+			),
+			canRemove: canRemoveBlocks( selectedBlockClientIds, rootClientId ),
+		};
+	}, [] );
+
+	if ( invalidBlocks ) {
+		return { isLoading: false, commands: [] };
+	}
 
 	const isTemplate = blocks.length === 1 && isTemplatePart( blocks[ 0 ] );
 ```

</p>
</details>

Reply:
> trick shallow equality check that the store has "fresh" 

Didn't think of that!

> but it might be easier just to return a flag value here and then bail out early in the command.

I will do as you suggest!

Thank you for context.

Reply:
Implemented!